### DISCUSSION TITLE
Failed Build openssl-3.0.5 for rbenv install 3.1.2

### DISCUSSION THREAD

**Discussion by cruzmanny** [rbenv/ruby-build#2053] [Build failures]
- I am on the latest version of Ubuntu 22.04.1
- Followed the instructions for the basic checkout git (also on a fresh install used the automated script)
- When running `rbenv install 3.1.2` an error message is received that the build failed because it is unable to make/install openssl-3.0.5. It seems openssl-3.0.5 is for the next version of Ubuntu (kinetic)
- After countless hours I was able to install ruby by downloading dependencies that are required to install openssl-3.0.5 and ruby. The below command will install the dependencies needed.

`sudo apt install libssl-dev libz-dev build-essential checkinstall zlib1y-dev`

-Perhaps the instructions or the script needs to be re-written?

**Comment by mislav** [comment]
(Transferred your issue to the ruby-build repository)

Hi, sorry you've had trouble, and thanks for sharing your solution steps!

If you had seen this earlier, would it have helped? https://github.com/rbenv/ruby-build/wiki#ubuntudebianmint
We could try to link to it more prominently from the `rbenv/rbenv` repository.

**Comment by cruzmanny** [comment]
Yes! That would have helped. I didn't see the checkinstall. It seems like openssl-3.0.5 requires it.

**Comment by JeffCohen** [comment]
(I hope it's ok to ask here but happy to move this to Discussion if that's preferred)

I regret posting because I know this has been a common issue, but I feel like I've read everything I can regarding workarounds and fear I'm missing something obvious. I'm also trying to install 3.1.2 on a new Ubuntu 22.04.1 LTS machine.  I made sure to `brew upgrade ruby-build` and also installed all of the extra extensions mentioned here and in the rbenv post that @mislav provided above.  I continue to get this error (I can post the full verbose log) if that's helpful.

My understanding is that ruby-build would try to grab the older version of openssl if needed, but I'm not seeing that happen (and maybe it doesn't need to in this case anyway?)

```
The Ruby openssl extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Try running `apt-get install -y libssl-dev` to fetch missing dependencies.

Configure options used:
  --prefix=/home/deploy/.rbenv/versions/3.1.2
  --enable-shared
  --with-readline-dir=/home/linuxbrew/.linuxbrew/opt/readline
  LDFLAGS=-L/home/deploy/.rbenv/versions/3.1.2/lib
  CPPFLAGS=-I/home/deploy/.rbenv/versions/3.1.2/include

BUILD FAILED (Ubuntu 22.04 using ruby-build 20220910.1)

Inspect or clean up the working tree at /tmp/ruby-build.20220923165340.2544.hqIt35
Results logged to /tmp/ruby-build.20220923165340.2544.log

Last 10 log lines:
The Ruby openssl extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Try running `apt-get install -y libssl-dev` to fetch missing dependencies.

Configure options used:
  --prefix=/home/deploy/.rbenv/versions/3.1.2
  --enable-shared
  --with-readline-dir=/home/linuxbrew/.linuxbrew/opt/readline
  LDFLAGS=-L/home/deploy/.rbenv/versions/3.1.2/lib
  CPPFLAGS=-I/home/deploy/.rbenv/versions/3.1.2/include
```

(Of course I've run `apt-get install -y libssl-dev` just to be triple-sure and still get this error. )

Any thoughts?

**Comment by mislav** [comment]
Could you paste the contents of the whole log file as a gist? Thanks

**Comment by JeffCohen** [comment]
Thanks mislav!  Here it is:

https://gist.github.com/JeffCohen/e701b7cedb08db06890c31cd79d91018

**Comment by mislav** [comment]
@JeffCohen Strange! This Ruby version should have been linked to your system OpenSSL, but for some reason it wasn't.

Can you check your environment variables to see if there is anything that could have affected compilation? See `env`

@hsbt Do you have any ideas?

**Comment by JeffCohen** [comment]
Here's my env vars: https://gist.github.com/JeffCohen/85339684ce96131d77318178cc986f03

Just in case it's important: this is my very first time using `brew` on Linux, honestly until this project I thought brew was only for mac.  I followed the standard install instructions and did nothing custom.  I used brew to install rbenv.  

But I used `apt` to install all the other dependencies per various instructions.  Maybe that's not ok because of certain path lookup rules (or something)?

**Comment by mislav** [comment]
Thanks for the share. I don't see anything odd in your env.

It shouldn't be a problem that you used Homebrew for anything. You should still use apt for system packages such as `libssl-dev`. And since you've installed OpenSSL headers, Ruby should have linked to them automatically, but for some reason it didn't. I also don't see a clear failure cause in the log that you shared.

In the meantime, you could try a workaround: install OpenSSL with Homebrew and instruct Ruby to link to that instead of the system one. Something like:

```sh
brew install openssl@3
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" rbenv install 3.1.2
```

**Reply by JeffCohen** [reply]
Thank you!  That worked perfectly.

In case it's a clue, when I installed openssl@3, brew said that it conflicts with 1.1 and therefore won't symlink it, and provided instructions for getting ver 3 earlier in the path that 1.1 if needed.

I'm not sure of the consequences of doing that yet, so I didn't do that but instead passed the command line arguments as you suggested, and 3.1.2 installed cleanly.

Thanks again.

**Reply by mislav** [reply]
Awesome; glad you solved it! I'm still puzzled that you had to do that though :(

> In case it's a clue, when I installed openssl@3, brew said that it conflicts with 1.1 and therefore won't symlink it, and provided instructions for getting ver 3 earlier in the path that 1.1 if needed.

Yeah that's a valid notice. When you install `openssl` with Homebrew it still doesn't take over as a default over the system one, since that could break other tools. It's best to use one-time installation configuration options like `RUBY_CONFIGURE_OPTS` to link to it instead of making it a system default.

But wait, you said that the message suggested that `openssl@1.1` was previously installed from Homebrew and symlinked as a system default? Do you remember doing that yourself? It could have been what botched your install in the first place. You can check with `brew info openssl@1.1`

**Reply by JeffCohen** [reply]
> It's best to use one-time installation configuration options like `RUBY_CONFIGURE_OPTS` to link to it instead of making it a system default.

Got it, thanks for that - makes total sense now.

The message from brew was that it detected that openssl 1.1 exists, even though I had not used brew to install it.  I may have installed 1.1 as part of a bunch of `apt install` commands when I first was setting up the new server.  

Just to be sure:

```
$ brew info openssl@1
Error: No available formula with the name "openssl@1". Did you mean openssl@1.1, openssl@3, openssh or openslp?
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.
```

The only commands I ran with brew are `brew install gcc` and then `brew install rbenv ruby-build`.  

Maybe `brew install gcc` installed a dependency that caused issues somehow?  At this stage it's hard for me to tell.

**Reply by mislav** [reply]
> Maybe `brew install gcc` installed a dependency that caused issues somehow? At this stage it's hard for me to tell.

Yes, hard to tell. BTW, on Ubuntu you should generally use `apt` to install compilers like gcc, for example with `sudo apt install build-essential` (if I remember the package name correctly). But maybe Homebrew for that is fine as well—I just haven't tried. 🤔

**Reply by hsbt** [reply]
Can you show the mkmf.log of openssl? Maybe it located at `'/tmp/ruby-build.20220923165340.2544.hqIt35/ruby-3.1.2/ext/openssl'`

And this issue may caused by linuxbrew. I think both openssl provided by linuxbrew/system are conflicted.

**Reply by huyz** [reply]
That `RUBY_CONFIGURE_OPTS` solved it for me too.

**Comment by JeffCohen** [comment]
Thanks @hsbt here's that log file: https://gist.github.com/JeffCohen/8f82c1f387f814ad251707d462631526

Maybe you can spot some clues?  I agree that linuxbrew is the prime suspect but can't see exactly why.

**Reply by mislav** [reply]
From the log it looks like it's linking to `/home/linuxbrew/.linuxbrew/Cellar/openssl@1.1/1.1.1q`

Which is wrong since Ruby 3.1 should never get linked to OpenSSL 1.1. On your Ubuntu, OpenSSL 3 should be the system default, but somehow Homebrew's OpenSSL 1.1 had precedence. I don't know why is this, but I suspect that you either allowed Homebrew's OpenSSL to be symlinked into PATH or that you've configured some environment variables somewhere to include it. Or, something else is amiss.

Ideally, `/home/linuxbrew/.linuxbrew/Cellar/openssl@1.1` should be entirely invisible to anything on your system until you explicitly choose to link something to it during compilation.

**Comment by JeffCohen** [comment]
Just in case anyone else gets this far, I've discovered that LinuxBrew dropped a file in my home folder named `ssl_info.txt`. This must have happened at some point and I never noticed it. 

Anyway here's the contents of this file, which, if I had read these instructions first, probably would have avoided my headaches with openssl 1.1 vs ruby 3.1:

```
openssl@3 is keg-only, which means it was not symlinked into /home/linuxbrew/.linuxbrew,
because it conflicts with the `openssl@1.1` formula.

If you need to have openssl@3 first in your PATH, run:
  echo 'export PATH="/home/linuxbrew/.linuxbrew/opt/openssl@3/bin:$PATH"' >> ~/.profile

For compilers to find openssl@3 you may need to set:
  export LDFLAGS="-L/home/linuxbrew/.linuxbrew/opt/openssl@3/lib"
  export CPPFLAGS="-I/home/linuxbrew/.linuxbrew/opt/openssl@3/include"

For pkg-config to find openssl@3 you may need to set:
  export PKG_CONFIG_PATH="/home/linuxbrew/.linuxbrew/opt/openssl@3/lib/pkgconfig"
```

**Comment by gisborne** [comment]
I am having this issue and nothing here helped. I have openssl 3.0.5 installed using homebrew. I've reinstalled it just now. From what I read, this command:

`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" rvm reinstall 3.1.2`

should have allowed me to install ruby 3.1.2 but it failed with the same error other folks are reporting:

ossl_ts.c:829:5: error: incomplete definition of type 'struct TS_verify_ctx'

In case it's helpful, 3.1.0 does the same thing.

**Reply by mislav** [reply]
Hi, this is an issue tracker for the ruby-build project (also accessed as `rbenv install` command). You seem to be using `rvm`, which is an entirely different Ruby version management tool, and you should ask RVM questions in a more appropriate forum. Thank you!

**Reply by yonibaciu** [reply]
For rvm the command is slightly different. You need to:

```
rvm install 3.1.2 --with-openssl-dir=$(brew --prefix openssl@3)
```

**Reply by combover01** [reply]
@yonibaciu I was having this problem on a Mac using rvm and your line solved all my problems!!! Thanks so much and I hope this helps others.
